### PR TITLE
Support TRL's RLOO

### PR DESCRIPTION
**Pull request by songxxzp**
As discussed in #515 , this is an RLOO implementation that follows TRL. Compared to TRL, OpenRLHF features a modern vllm inference engine, making it a more efficient reinforcement learning implementation. Therefore, this Pull Request aims to provide users who wish to use the TRL RLOO algorithm with a more efficient option.
Screenshot_20241204_193850
[openrlhf_train_ppo](https://wandb.ai/songxxzp/openrlhf_train_ppo?nw=nwusersongxxzp)

It should be noted that, since RLOO treats an entire response as a single sentence, the shape of its rewards and advantages is [batch size, 1], which differs from the [batch size, seq len] shape of advantages in PPO. To maintain compatibility with the framework, in this implementation, after computing the advantages, they are placed into a zero tensor of shape [batch size, seq len], with the insertion point being the end of the sentence (determined by the action mask).

Another difference is that RLOO's advantages do not require normalization across the batch. When calculating advantages, RLOO inherently performs a certain degree of normalization.

Finally, when computing RLOO's loss, it is necessary to use floats and apply clipping in advance to prevent numerical overflow.

Tested with following setting:
```
openrlhf.cli.train_ppo \
   --pretrain OpenRLHF/Llama-3-8b-sft-mixture \
   --reward_pretrain OpenRLHF/Llama-3-8b-rm-mixture \
   --save_path ./checkpoint/llama-3-8b-rloo-128-128x8 \
   --save_steps -1 \
   --logging_steps 1 \
   --eval_steps -1 \
   --micro_train_batch_size 2 \
   --train_batch_size 64 \
   --micro_rollout_batch_size 8 \
   --rollout_batch_size 8 \
   --max_epochs 1 \
   --prompt_max_len 1024 \
   --generate_max_len 1024 \
   --zero_stage 2 \
   --bf16 \
   --actor_learning_rate 5e-7 \
   --critic_learning_rate 9e-6 \
   --init_kl_coef 0.001 \
   --prompt_data OpenRLHF/prompt-collection-v0.1 \
   --input_key context_messages \
   --apply_chat_template \
   --max_samples 25000 \
   --normalize_reward \
   --adam_offload \
   --flash_attn \
   --load_checkpoint \
   --gradient_checkpointing \
   --advantage_estimator trl_rloo \
   --n_samples_per_prompt 8 \
   --wandb_run_name rloo_64-8x8_1203_04
```

### CODE REVIEW

**Comment by hijkzzz** [comment]
Please compare it with REINFORCE using the same hyper-parameters.

**Comment by songxxzp** [comment]
Thank you for the suggestion. The rollout batch size is set relatively small due to limited computing resources. I will try to scale up soon.

**Comment by songxxzp** [comment]
@hijkzzz Hi, here come the results:
Screenshot_20241209_171728
[openrlhf_train_ppo](https://wandb.ai/songxxzp/openrlhf_train_ppo?nw=nwusersongxxzp)
The return of rloo with per token KL(orl_rloo) looks different, which makes me curious.
Additionally, previously setting init_kl_coef=0.001 https://github.com/OpenRLHF/OpenRLHF/pull/515#issuecomment-2514353697 caused the KL to spike, rendering the experimental results meaningless, so I used 0.01 instead.
The configurations only differ in the advantage estimator used:
```bash
openrlhf.cli.train_ppo \
   --pretrain OpenRLHF/Llama-3-8b-sft-mixture \
   --reward_pretrain OpenRLHF/Llama-3-8b-rm-mixture \
   --save_path /workspace/mm_data/video_reward_ckpt/OpenRLHF/checkpoint/llama-3-8b-reinforce-512-64x8_kl_0.01 \
   --ckpt_path /workspace/mm_data/video_reward_ckpt/OpenRLHF/checkpoint/llama-3-8b-reinforce-512-64x8_kl_0.01 \
   --save_steps 100 \
   --max_ckpt_num 1 \
   --logging_steps 1 \
   --eval_steps -1 \
   --micro_train_batch_size 2 \
   --train_batch_size 512 \
   --micro_rollout_batch_size 8 \
   --rollout_batch_size 64 \
   --max_epochs 1 \
   --prompt_max_len 1024 \
   --generate_max_len 1024 \
   --advantage_estimator reinforce \
   --zero_stage 2 \
   --bf16 \
   --actor_learning_rate 5e-7 \
   --critic_learning_rate 9e-6 \
   --init_kl_coef 0.01 \
   --prompt_data OpenRLHF/prompt-collection-v0.1 \
   --input_key context_messages \
   --apply_chat_template \
   --max_samples 6096 \
   --normalize_reward \
   --adam_offload \
   --flash_attn \
   --gradient_checkpointing \
   --n_samples_per_prompt 8 \
   --freezing_actor_steps 0 \
   --use_wandb 18b5599a6315de08ffd4e13faa1f0f039e52b6b3 \
   --wandb_run_name reinforce_512-64x8_kl_0.01_1206_02
```
